src/framework/global/log.h

```diff
@@ -31,6 +31,7 @@
 #define DO_ASSERT_X(cond, msg) \
     LOGE() << "\"ASSERT FAILED!\": " << msg << ", file: " << __FILE__ << ", line: " << __LINE__; \
     assert(cond); \
+    (void)(cond); \    // avoid compiler warning
```

Comment:
In a release build, doesn't this mean that the condition is unnecessarily evaluated? 

If you have `assert(someSimpleBooleanVariable)`, it's fine. But if you have `assert(evaluate(some, complicated[expression]))`, then I'm afraid that it means that `evaluate(some, complicated[expression])` will be evaluated unnecessary. As far as I know, `(void)` only discards the result, but does not per definition mean that the expression won't be evaluated. 

As a solution, I propose to make two macros: 
- `DO_ASSERT` which is simply `assert`, so that it can be used for complicated expressions and does completely nothing in release builds
- `ASSERT_OR_UNUSED` which is useful for asserting the value of a variable that would otherwise be unused.

Reply:
I made the first implementation a little careless
Changed the implementation:
* Now the assertion message will be logged in the release as well. Of course, the expression will also be evaluated in the release, but I think the assertion triggering is a mistake, we should not hide it.
* If we need to check something only in debug, I added an explicit macro `DO_ASSERT_IF_DEBUG`

What do you think of this approach?

Reply:
That makes sense to me. Seeing the failure message in release builds may help understanding crashes from log files submitted by users. 

However, of course, we must make sure that for every case we make a wise decision about whether to use `DO_ASSERT` or `DO_ASSERT_ON_DEBUG`; for cheap assertions that check a critical error condition, we should use `DO_ASSERT`, but for more expensive assertions that are actually just "sanity checks", we should consider using `DO_ASSERT_ON_DEBUG` instead.